main.py

```diff
@@ -175,90 +282,72 @@ def wrapped_main(options: Options, heartbeat: Callable[[], None]) -> int:
-        timings = ''
-        if options.show_timings:
-            time_perm += t1 - t0
-            time_compile += t2 - t1
-            time_score += t3 - t2
-            time_total = time_perm + time_compile + time_score
-            timings = ", {}% perm, {}% compile, {}% score".format(
-                round(100 * time_perm / time_total),
-                round(100 * time_compile / time_total),
-                round(100 * time_score / time_total)
-            )
-        status_line = f"iteration {iteration}, {errors} errors, score = {disp_score}{timings}"
-
-        if new_score <= perm.base_score and new_hash and new_hash not in perm.hashes:
-            perm.hashes.add(new_hash)
-            print("\r" + " " * (len(status_line) + 10) + "\r", end='')
-            if new_score < perm.base_score:
-                high_scores[perm_ind] = new_score
-                print(f"[{perm.unique_name}] found a better score! ({new_score} vs {perm.base_score})")
-            else:
-                print(f"[{perm.unique_name}] found different asm with same score")
-
-            source = perm.get_source()
-            write_candidate(perm, source)
-        print("\b"*10 + " "*10 + "\r" + status_line, end='', flush=True)
-    
-    return min(high_scores)
+    for permuter in context.permuters:
+        if name_counts[permuter.fn_name] > 1:
+            permuter.unique_name += f" ({permuter.dir})"
+        print(f"[{permuter.unique_name}] base score = {permuter.base.score_value}")
+
+    context.high_scores = dict([(p.base.score_value, p) for p in context.permuters])
+
+    perm_seed_iter = gen_all_seeds(context.permuters, heartbeat)
+    if options.threads == 1:
+        for permuter_index, seed in perm_seed_iter:
+            heartbeat()
+            permuter = context.permuters[permuter_index]
+            # Run single threaded
+            exception = None
+            try:
+                cand, profiler = permuter.eval_candidate(seed)
+            except Exception as e:
+                cand = None
+                profiler = None
+                exception = e
+            post_score(context, permuter, cand, exception, profiler)
+    else: # Run multi-threaded
+        # Create queues
+        task_queue = Queue()
+        results_queue = Queue()
+
+        def worker(input, output):
+            global context
+            for permuter_index, seed in iter(input.get, 'STOP'):
+                try:
+                    permuter = context.permuters[permuter_index]
+                    result = permuter.eval_candidate(seed) + (permuter_index,None,)
+                except e:
+                    result = (None, None, permuter_index, e)
+                output.put(result)
+
+        # Begin workers
+        for i in range(options.threads):
+            Process(target=worker, args=(task_queue, results_queue)).start()
+        
+        # Run batch work
+        for bat in iter_util.batch(iter(perm_seed_iter), options.threads * 100):
```

Comment:
any reason to do batching with the persistent worker setup? seems like you could just as well put #worker * 2 or so tasks in the queue initially, and every time an item is finished, put in the next. Something like:

```
active_items = 0
q = Queue()
def more_work():
  nonlocal active_items, q
  work = next(iter, None)
  if work is not None:
    q.put(work)
    active_items += 1

for i in range(#workers * 2):
  more_work()

while active_items > 0:
  heartbeat()
  # (wait for one)
  active_items -= 1
  more_work()
```